cli/src/complete.rs

```diff
@@ -271,6 +271,73 @@ pub fn operations() -> Vec<CompletionCandidate> {
     })
 }
 
+fn config_keys_rec(
+    prefix: &str,
+    properties: &serde_json::Map<String, serde_json::Value>,
+    acc: &mut Vec<CompletionCandidate>,
+    only_leaves: bool,
+) {
+    for (key, value) in properties {
+        if key == "type" {
+            // used by "additionalProperties", ignore
+            continue;
+        }
+        let owned_key;
+        let key = if key.contains('(') {
+            // Parenthesis can occur in revset aliases, e.g. immutable_heads().
+            // These must be escaped.
+            owned_key = format!("'{key}'");
+            &owned_key
+        } else {
+            key
+        };
+        let value = value.as_object().unwrap();
+        match value.get("type").and_then(|v| v.as_str()) {
+            Some("object") => {
+                if !only_leaves {
+                    acc.push(CompletionCandidate::new(format!("{prefix}{key}")));
+                }
+                let Some(properties) = value.get("properties") else {
+                    continue;
+                };
+                let properties = properties.as_object().unwrap();
+                config_keys_rec(&format!("{prefix}{key}."), properties, acc, only_leaves);
+            }
+            _ => {
+                let help = value
+                    .get("description")
+                    .map(|desc| desc.as_str().unwrap().to_string().into());
+                acc.push(CompletionCandidate::new(format!("{prefix}{key}")).help(help));
+            }
+        }
+    }
+}
+
+fn config_keys_impl(only_leaves: bool) -> Vec<CompletionCandidate> {
+    let schema = include_str!("config-schema.json");
```

Comment:
nit: better to extract the constant to the config module than including blob at two places.